routing/index_graph.hpp

```diff
@@ -25,36 +33,159 @@ class JointEdge final
+  }
+
+  Joint::Id GetFrom() const { return m_from; }
+  Joint::Id GetTo() const { return m_to; }
+  uint32_t GetFeatureId() const { return m_featureId; }
+
+private:
+  Joint::Id const m_from = Joint::kInvalidId;
+  Joint::Id const m_to = Joint::kInvalidId;
+  // Note. It's important to store feature id because two |m_from| and |m_to| may be
+  // connected with several features.
+  uint32_t const m_featureId = 0;
+};
+
+inline bool IsAdjacent(DirectedEdge const & ingoing, DirectedEdge const & outgoing)
+{
+  return ingoing.GetTo() == outgoing.GetFrom();
+}
+
+class RestrictionInfo final
+{
+public:  
+  RestrictionInfo() = default;
+
+  RestrictionInfo(DirectedEdge const & ingoing, DirectedEdge const & outgoing)
+    : m_from(ingoing.GetFrom()), m_center(ingoing.GetTo()), m_to(outgoing.GetTo()),
+      m_fromFeatureId(ingoing.GetFeatureId()), m_toFeatureId(outgoing.GetFeatureId())
+  {
+    CHECK(IsAdjacent(ingoing, outgoing), ());
+  }
+
+  bool operator<(RestrictionInfo const & rhs) const
+  {
+    if (m_from != rhs.m_from)
+      return m_from < rhs.m_from;
+
+    if (m_center != rhs.m_center)
+      return m_center < rhs.m_center;
+
+    if (m_to != rhs.m_to)
+      return m_to < rhs.m_to;
+
+    if (m_fromFeatureId != rhs.m_fromFeatureId)
+      return m_fromFeatureId < rhs.m_fromFeatureId;
+
+    return m_toFeatureId < rhs.m_toFeatureId;
+  }
+
+  bool operator==(RestrictionInfo const & rhs) const { return !(*this < rhs || rhs < *this); }
+
+  bool operator!=(RestrictionInfo const & rhs) const { return !(*this == rhs); }
+
+  pair<DirectedEdge, DirectedEdge> Edges()
+  {
+    return make_pair(DirectedEdge(m_from, m_center, m_fromFeatureId),
+                     DirectedEdge(m_center, m_to, m_toFeatureId));
+  }
+
+  Joint::Id m_from = Joint::kInvalidId;
+  Joint::Id m_center = Joint::kInvalidId;
+  Joint::Id m_to = Joint::kInvalidId;
+  uint32_t m_fromFeatureId = 0;
+  uint32_t m_toFeatureId = 0;
 };
 
 class IndexGraph final
 {
 public:
+  friend struct routing_test::RestrictionTest;
+
+  static uint32_t const kStartFakeFeatureIds = 1024 * 1024 * 1024;
+
   IndexGraph() = default;
   explicit IndexGraph(unique_ptr<GeometryLoader> loader, shared_ptr<EdgeEstimator> estimator);
 
   // Creates edge for points in same feature.
   void GetDirectedEdge(uint32_t featureId, uint32_t pointFrom, uint32_t pointTo, Joint::Id target,
                        bool forward, vector<JointEdge> & edges);
-  void GetNeighboringEdges(RoadPoint const & rp, bool isOutgoing, vector<JointEdge> & edges);
+  void GetNeighboringEdges(RoadPoint const & rp, bool isOutgoing, bool graphWithoutRestrictions,
+                           vector<JointEdge> & edges);
+  /// \brief Fills |edges| with edges go through |rp|. |rp| should not be an joint.
+  void GetIntermediatePointEdges(RoadPoint const & rp, bool graphWithoutRestrictions,
+                                 vector<DirectedEdge> & edges);
 
   // Put outgoing (or ingoing) egdes for jointId to the 'edges' vector.
```

Comment:
I didn't offer as I'm not a salesman. I proposed to write ```|edges|``` instead of ``'edges'``.